src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/ConcatenationRewriter.cs

```diff
@@ -0,0 +1,76 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System.Collections.Generic;
+using EnsureThat;
+using Microsoft.Health.Fhir.Core.Features.Search.Expressions;
+
+namespace Microsoft.Health.Fhir.SqlServer.Features.Search.Expressions.Visitors
+{
+    /// <summary>
+    /// An abstract rewriter to rewrite a table expression into a concatenation of two table expressions.
+    /// </summary>
+    internal abstract class ConcatenationRewriter : SqlExpressionRewriterWithDefaultInitialContext<object>
+    {
+        private readonly IExpressionVisitor<object, bool> _scout;
+
+        protected ConcatenationRewriter(IExpressionVisitor<object, bool> scout)
+        {
+            EnsureArg.IsNotNull(scout, nameof(scout));
+            _scout = scout;
+        }
+
+        public override Expression VisitSqlRoot(SqlRootExpression expression, object context)
+        {
+            if (expression.TableExpressions.Count == 0)
+            {
+                return expression;
+            }
+
+            List<TableExpression> newTableExpressions = null;
+            for (var i = 0; i < expression.TableExpressions.Count; i++)
+            {
+                TableExpression tableExpression = expression.TableExpressions[i];
+
+                if (tableExpression.NormalizedPredicate.AcceptVisitor(_scout, null))
+                {
+                    if (newTableExpressions == null)
+                    {
+                        newTableExpressions = new List<TableExpression>();
```

Comment:
nit: `newTableExpressions = expression.TableExpression.Take(i).ToList()`

Reply:
This is extracted out to a method in #527.